expression/builtin_info.go

```diff
@@ -771,6 +776,122 @@ func (k TiDBDecodeKeyFunctionKeyType) String() string {
 // TiDBDecodeKeyFunctionKey is used to identify the decoder function in context.
 const TiDBDecodeKeyFunctionKey TiDBDecodeKeyFunctionKeyType = 0
 
+type tidbDecodeSQLDigestsFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *tidbDecodeSQLDigestsFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+
+	var argTps []types.EvalType
+	if len(args) > 1 {
+		argTps = []types.EvalType{types.ETString, types.ETInt}
+	} else {
+		argTps = []types.EvalType{types.ETString}
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, argTps...)
+	if err != nil {
+		return nil, err
+	}
+	sig := &builtinTiDBDecodeSQLDigestsSig{bf}
+	return sig, nil
+}
+
+type builtinTiDBDecodeSQLDigestsSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinTiDBDecodeSQLDigestsSig) Clone() builtinFunc {
+	newSig := &builtinTiDBDecodeSQLDigestsSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinTiDBDecodeSQLDigestsSig) evalString(row chunk.Row) (string, bool, error) {
+	args := b.getArgs()
+	digestsStr, isNull, err := args[0].EvalString(b.ctx, row)
+	if err != nil {
+		return "", true, err
+	}
+	if isNull {
+		return "", true, nil
+	}
+
+	stmtTruncateLength := int64(0)
+	if len(args) > 1 {
+		stmtTruncateLength, isNull, err = args[1].EvalInt(b.ctx, row)
+		if err != nil {
+			return "", true, err
+		}
+		if isNull {
+			stmtTruncateLength = 0
+		}
+	}
+
+	var digests []interface{}
+	err = json.Unmarshal([]byte(digestsStr), &digests)
+	if err != nil {
+		const errMsgMaxLength = 32
+		if len(digestsStr) > errMsgMaxLength {
+			digestsStr = digestsStr[:errMsgMaxLength] + "..."
+		}
+		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errIncorrectArgs.GenWithStack("The argument can't be unmarshalled as JSON array: '%s'", digestsStr))
+		return "", true, nil
+	}
+
+	// Query the SQL Statements by digests.
+	retriever := NewSQLDigestTextRetriever()
+	for _, item := range digests {
+		if item != nil {
+			digest, ok := item.(string)
+			if ok {
+				retriever.SQLDigestsMap[digest] = ""
+			}
+		}
+	}
+
+	// Querying may take some time and it takes a context.Context as argument, which is not available here.
+	// We simply create a context with 1s timeout here.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```

Comment:
Is 1s enough?

Reply:
I'm not sure, and I set it short because I'm afraid if there are too many ongoing transactions, `select tidb_decode_sql_digests(all_sql_digests) from cluster_tidb_trx` will take too long time since it does this internal query for each row in cluster_tidb_trx. But perhaps it would be worse if the user's query fails. How do you think? How about 3s or 5s?

Reply:
I'm actually not sure if here's not the proper place to implement this internal query. The fact that the `context.Context` is not present here may indicate that a builtin function's execution is expected to be fast and light weighted. However I didn't find any better way to implement this function.

Reply:
I don't think it's suitable for users to do `select tidb_decode_sql_digests(all_sql_digests) from cluster_tidb_trx`. In my expectation, users only do `select tidb_decode_sql_digests("digests")` like what they do to decode plan digest. If we change the `all_sql_digests` format to `tidb_decode_sql_digests("digests")` not `digests`, I think we can avoid such a case, but the disadvantage is it's troublesome to decode multiple transactions' digests.

Reply:
I didn't understand. Did you mean making the tidb_trx table show the decoded sql array directly instead of the current `all_sql_digests` column?

Reply:
I think we can tell users not to do this in the documents, and it should be seldom used by users, so we can spend less energy on it.

Reply:
@cfzjywxk I'm afraid if there are multi rows, each row will have this timeout. And `max_execution_time` can be unlimited. How about using max(max_execution_time, 20s)?

Reply:
I think it's better to let the user control the resources and execution limits if possible. As the inner restricted sql executor is used, the unlimited `max_execution_time` is not very suited either. Also we don't want to add more configurations or variables, maybe we could add a implicit limit first, `20s` seems to be ok.

Reply:
@cfzjywxk Done in last commit PTAL again, thanks!